rust/operator-binary/src/controller.rs

```diff
@@ -174,15 +174,16 @@ pub async fn reconcile_hive(hive: Arc<HiveCluster>, ctx: Arc<Ctx>) -> Result<Act
     let resolved_product_image: ResolvedProductImage =
         hive.spec.image.resolve(DOCKER_IMAGE_BASE_NAME);
 
-    let s3_connection_spec: Option<S3ConnectionSpec> = if let Some(s3) = &hive.spec.s3 {
-        Some(
-            s3.resolve(client, hive.namespace().as_deref().unwrap())
-                .await
-                .context(ResolveS3ConnectionSnafu)?,
-        )
-    } else {
-        None
-    };
+    let s3_connection_spec: Option<S3ConnectionSpec> =
+        if let Some(s3) = &hive.spec.cluster_config.s3 {
+            Some(
+                s3.resolve(client, hive.namespace().as_deref().unwrap())
+                    .await
+                    .context(ResolveS3ConnectionSnafu)?,
```

Comment:
New error variant (I suggest to put it at top)
```
    #[snafu(display("object defines no namespace"))]
    ObjectHasNoNamespace,
```

```suggestion
                s3.resolve(
                    client,
                    &hive.namespace().ok_or(Error::ObjectHasNoNamespace)?,
                )
                .await
                .context(ResolveS3ConnectionSnafu)?,
```